### PR TITLE
Translation/rotation code for a 4D series of volumes

### PR DESCRIPTION
**Pull request by jcbmiller94**
Please note: this code is still flawed, as noted in the presentation yesterday. I am making this pull request to share the code with everyone for testing, and to keep documentation before I implement the planned changes below:

- optimizing all 6 rigid body transformations at once (instead of translations followed by rotations)
- adding random noise to the voxel coordinates, as to better capture shifts between volumes that are smaller than the size of a voxel grid shift

### CODE REVIEW

**Comment by matthew-brett** [comment]
Hey - sorry not to get back to you immediately on the nipy code to add a tiny bit of coordinate jitter when resampling the images.

Here's the only mention I could find of this in the SPM12 code at the moment: https://bitbucket.org/matthewbrett/spm12/src/ec9cae02ebf8bb367ba71e2782e2be1a77a67e28/toolbox/OldNorm/spm_affreg.m?at=master&fileviewer=file-view-default#spm_affreg.m-94

It occurred to me that it would not be too hard for you to do the same thing yourself.  Have a look at https://bic-berkeley.github.io/psych-214-fall-2016/map_coordinates.html . Your job would be to create an explicit coordinate grid for the image you are registering to - as in that page - and then add some random jitter to the coordinates in the grid, as in the SPM code.   Then do the affine transformation on the coordinates, and use the scipy `map_coordinates` function to sample the moving image, instead of ``affine_transform``.    You could check whether the cost function really is smoother when you do that.

**Comment by jcbmiller94** [comment]
Thanks for this! I haven't looked at the SPM code yet, as I am currently changing the framework to work on optimizing the rigid body transforms for the 6 parameters at once. I am doing this the the `map_coordinates` function, but I'm wondering about how `affine_transform` plays a role. Why is it that I should do an affine transform on the coordinates before the general resampling with the coordinate mapping? Or is the affine transform just for when the random noise is added. (Not sure if this question makes sense. I should have my code pushed up later tonight, and can point you to my area of confusion there)

**Comment by jcbmiller94** [comment]
I wrote the following two new scripts (and with proper Python format documentation parameters!) 

This is my first attempt to use the scipy `map_coordinates` function approach of general image resampling to capture rigid body motion parameters between volumes. I also wrote a test, which failed, given the following inputs and outputs of translation and rotation values. 

testing parameters: [-1.  , -3.  , -2.  ,  0.5 ,  0.07, -0.3 ]
'optimized' output: [[ -1.2382e+01,   1.5069e+01,  -1.1921e+01,  -1.6186e-03, 2.2384e-05,  -4.3019e-02]

@matthew-brett this seems as a case of getting caught in local minima? Given the cost function went through four iterations with a final value of -.52 or so. Assuming my framework is set up and the functions are operating as they should, I think it's time for me to investigate the SPM noise adding algorithm

**Comment by jcbmiller94** [comment]
@zzbalews @mpnagle This took a lot of looking at your code!! It was really written well and that made it wayyyy easier to try and use a similar process here. Maybe on Friday I can show you all this and get your input of what you think might be happening? Or if you think this works?

**Comment by zzbalews** [comment]
Cool I'm glad our code helped! Have you tried using bigger translations on this test example? I applied the the affine you're using (img on left) and then the inverse to put it back (img on right). (This is ds107* but I think it'll look similar for ds114). A big chunk of the brain is missing, so maybe your best parameters aren't matching because voxels that should be part of the brain are 0s.

screen shot 2016-12-08 at 12 07 31 am

**Comment by jcbmiller94** [comment]
Hey @zzbalews  - just to summarize our discussion from earlier today and have it documented on here. Using my test rotation and translation parameters, the brain, as you said, moves off of the voxel grid and is unrecoverable with the transforms we've been using. Taking your advice, I applied a rotation lower in magnitude, and recovered it with the inverse of the affine I transformed it with (see picture). Now, I should be confident that this rotation is appropriate for testing the map coordinates code. 

image

**Comment by jcbmiller94** [comment]
image

(and even more obvious example, with bigger translations)

File: code/fmri_utils/func_preproc/optimize_map_coordinates.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    # affine maping bet the moving affine and the affine of the reference volume
+    vol1_to_ref_vol = npl.inv(moving_affine).dot(ref_vol_affine)
+
+    # get shape of reference volume
+    I, J, K = ref_vol.shape
+    # create a mesh coordinate grid
+    i_vals, j_vals, k_vals = np.meshgrid(range(I), range(J), range(K), indexing='ij')
+    in_vox_coords = np.array([i_vals, j_vals, k_vals])
+
+    ### Add SPM random noise algorithm here to coordinates !!
```
Comment (matthew-brett) [inline code/fmri_utils/func_preproc/optimize_map_coordinates.py]: Right - here is where you could add some jitter to the `in_vox_coords` - why not try random numbers between -0.5 and 0.5 to each of the i, j, k coordinates?  You could also try and work out how to create this jittered grid before starting the registration and use the same jittered grid for all the volumes.

**Comment by jcbmiller94** [comment]
Thanks @matthew-brett! Attempted to implement this in the code I just pushed. Below are the parameters I get out of running the same data through SPM and my own code. The good news is that the values are no longer all at 0, however, the drift that SPM is observing is not generally seen in this code. 

screen shot 2016-12-04 at 10 34 58 pm

rp_map_coords_sub-10159_task-rest_bold

Do you have an idea of how I could check the smoothness of the cost function? I will run this with and without the jitter added, as I'm wondering if the fluctuations are capturing actual brain movement, or jsut the noise that was added to the coordinate grid

**Comment by zzbalews** [comment]
@jcbmiller94 Woohoo! Nice plots :) If it's not too much work, do you think it would change anything to initialize each optimization with the best parameters from the preceding volume? (You're using all 0s now, right?) I think that would be a reasonable first guess (unless the participant is very very jittery), and it'll probably speed up search time and help avoid local minima in case you're getting stuck.

**Comment by jcbmiller94** [comment]
image

@zzbalews this is what happens when I use the previous volume's parameters as the best guess. I'm still worried that all I'm capturing is the jitter, so I'll re-run this 6 parameter approach without the added noise and report back!

File: code/fmri_utils/func_preproc/optimize_map_coordinates.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    vol1_to_ref_vol = npl.inv(moving_affine).dot(ref_vol_affine)
+
+    # get shape of reference volume
+    I, J, K = ref_vol.shape
+    # create a mesh coordinate grid
+    i_vals, j_vals, k_vals = np.meshgrid(range(I), range(J), range(K), indexing='ij')
+    in_vox_coords = np.array([i_vals, j_vals, k_vals])
+
+    #create array of 3*(I,J,K) random numbers between -0.5 and 0.5
+    jitter = np.random.uniform(-0.5,0.5, 3*I*J*K)
```
Comment (matthew-brett) [inline code/fmri_utils/func_preproc/optimize_map_coordinates.py]: Or `jitter = np.random.uniform(-0.5,0.5, size=in_vox_coords.shape)`.

Why not try:

* turning down the jitter (-0.1 to 0.1)
* using a single jitter to realign all the volumes in the run?

**Comment by jcbmiller94** [comment]
The images below are 
- (1) reduced jitter to -0.1 to 0.1, using a random seed
- (2) same parameters as (1), but using the previous volume's realignment parameters as the best guess for the next one 

image

image

@matthew-brett you mentioned using smoothing to realign the images, but then returning a non-smoothed images. My question for this would be, at what point would I initiate the smoothing for each volume - right before the `map_coordinates` step? And how would I then 'unsmooth' the volume?

**Comment by matthew-brett** [comment]
I only meant that you smooth the volumes to calculate the realignment, but then you use the unsmoothed volumes to resample the images at the end.

File: code/fmri_utils/func_preproc/volume_realignment.py [hunk 3]
```diff
@@ -51,7 +51,8 @@ def volume_4D_realign(img_path, reference = 0, smooth_fwhm = 0, prefix = 'r', dr
     # smooth image, if designated
     if smooth_fwhm > 0:
         fwhm = smooth_fwhm
-        img = proc.smooth_image(img, 5)
+        img_smooth = proc.smooth_image(img, fwhm)
+        data_smooth = img.get_data()
```
Comment (ctseng12) [inline code/fmri_utils/func_preproc/volume_realignment.py]: Do you mean `data_smooth = img_smooth.get_data()` here?

Comment (jcbmiller94) [reply code/fmri_utils/func_preproc/volume_realignment.py]: Yes, that is exactly what I meant :) thank you!!!

**Comment by jcbmiller94** [comment]
screen shot 2016-12-13 at 1 51 58 pm

Results when using the smoothed data - still the same questions. I emailed @matthew-brett about meeting to go over the SPM code, hopefully try to understand the differences better

**Comment by zzbalews** [comment]
what happens when you do the smoothing and random jitter, but always initialize at 0 (not the previous output)? (or also skip the jitter?) Maybe now that the data is smoothed the noise you're introducing is stronger than signal from small movements in your images. Could that explain the really correlated spiky traces?

Also, are you always using the same random jitter on purpose?

**Comment by jcbmiller94** [comment]
@zzbalews trying this now! It was Matthew's idea to use the same random jitter -  I figure it's ok to use consistent noise because this is being done between two volumes at a time, and for each one the results might be less prone to randomness. But, I can test this. 

Other things I was going to test out in the next 2 days:
 - using the middle volume or a mean volume as a reference image
- using a multi-pass system with smoothing where initially 8mm, then 4mm, then 2mm, then no smoothing is applied to the data before getting the realignment parameters

File: code/fmri_utils/func_preproc/optimize_map_coordinates.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    # affine maping bet the moving affine and the affine of the reference volume
+    vol1_to_ref_vol = npl.inv(moving_affine).dot(ref_vol_affine)
+
+    # get shape of reference volume
+    I, J, K = ref_vol.shape
+    # create a mesh coordinate grid
+    i_vals, j_vals, k_vals = np.meshgrid(range(I), range(J), range(K), indexing='ij')
+    in_vox_coords = np.array([i_vals, j_vals, k_vals])
+
+    np.random.seed(1500)
```
Comment (zzbalews) [inline code/fmri_utils/func_preproc/optimize_map_coordinates.py]: Good point about trying to get the same results. What if you move `np.random.seed(1500)` outside this function/to the top of your script? That way the jitter will vary for every pair of volumes, but it'll be the same every time you run it.

Comment (jcbmiller94) [reply code/fmri_utils/func_preproc/optimize_map_coordinates.py]: Yeah, that's a great middle ground idea! Will try before the next two ideas.

**Comment by jcbmiller94** [comment]
image

This is with: 
- jitter the exact same for each volume 
- guess parameters always at [0,0,0,0,0,0] 
- 5mm smoothing

If you look at the z translation (red) it actually looks quite similar in overall shape to the SPM plot! But the other parameters, especially rotations, still look a little off (although here, there are less correlated spiking)

**Comment by matthew-brett** [comment]
Jacob - I remembered last night - be careful to remove the areas in the image equal to zero from the cost function calculation.   Zeros result when the movement causes the resampling to sample outside the image that is moving - so returning a default value, 0.  You need to make sure to remove these voxels with zero from the mismatch calculation, otherwise the correlation (etc) will drop just because there are zeros at the edge of the brain - and therefore penalize especially z translation, x, y rotation.

**Comment by jcbmiller94** [comment]
@matthew-brett I tried implementing a non-zero filter on the reference and moving volume. However, when I do this and then correlate them in the mismatch function, I am getting an error because the vectors are now of different length, and the correlation cannot be computed (using `np.corrcoeff`). 

Could a cross-correlation be useful here? (https://docs.scipy.org/doc/scipy-0.18.1/reference/generated/scipy.signal.correlate.html) Or, is there a way to preserve the shapes by, for example, using NaNs as replacements of the zeros

**Comment by matthew-brett** [comment]
The trick is to find all the voxels in the resampled moving image that are now zero because of sampling outside the volume, and drop those voxels from both images, before calculating the cost function.  I guess you were only dropping the zeros from the moving image?

**Comment by jcbmiller94** [comment]
Tried to do implement this in the cost function with my latest commit. It's currently running, and I'll see how the results look

**Comment by jcbmiller94** [comment]
image

This is with:

- jitter the exact same for each volume
- guess parameters always at [0,0,0,0,0,0]
- 5mm smoothing
- dropping voxels that have a value of 0 from the resampling of the moving image

Visually, looks like smaller changes are being caught. I'll report back on using a two-pass system tonight. But the complexity of SPM realignment, I think, is what's making the results different. Their code uses cost functions and optimization methods that use first and second derivatives, as well, and are better at getting out of local minima

**Comment by mpnagle** [comment]
😺 😺 😺

**Pull request merged**